### PR TITLE
Add infra for optimization visualization and benchmarking

### PR DESCRIPTION
**Pull request by ProfFan**
Per #81, I created a solution with PythonKit-wrapped `tensorboardX` to generate the optimization logs:

image

How to run:
```
mkdir -p logs/lm_sphere
swift build -c release -Xswiftc -cross-module-optimization
.build/release/Pose3SLAMG2O ./sphere2500.g2o ./result_pose3.txt ./logs/lm_sphere

# in a separate terminal
tensorboard dev upload --logdir logs \
    --name "First SwiftFusion in tensorboard.dev" \
    --description "One small step for myself, a giant step for human kind."
```

Demo:
https://tensorboard.dev/experiment/acL9bYbbQJKi2mAyPZoYDw/#scalars&runSelectionState=eyIxIjpmYWxzZSwiZ3JpZDNkIjpmYWxzZSwiZ2FyYWdlIjpmYWxzZSwic3BoZXJlX2JpZ25vaXNlIjpmYWxzZSwidGVzdF9mcm9iZW5pdXMiOmZhbHNlLCJ0ZXN0X2Zyb2Jlbml1c19maXhlZCI6ZmFsc2UsInRvcnVzIjpmYWxzZX0%3D

### CODE REVIEW

**Comment by dellaert** [comment]
That’s cool!

File: Examples/Pose3SLAMG2O/main.swift [hunk 1]
```diff
@@ -45,10 +46,13 @@ struct FileHandlerOutputStream: TextOutputStream {
 
 func main() {
   // Parse commandline.
-  guard CommandLine.arguments.count == 3 else {
+  guard CommandLine.arguments.count > 3 else {
```
Comment (dellaert) [inline Examples/Pose3SLAMG2O/main.swift:49]: Is there not a google flags like library that you can use?

Comment (ProfFan) [reply Examples/Pose3SLAMG2O/main.swift:49]: I hacked this through to avoid adding dependencies, can definitely improve later

Comment (dellaert) [reply Examples/Pose3SLAMG2O/main.swift:49]: Nah, I think you should do it now :-)

**Comment by ProfFan** [comment]
Probably I did something wrong in implementing the Chordal loss, but currently the chordal loss result looks very strange:
image

File: Examples/Pose3SLAMG2O/main.swift (hunk 1 above)
Comment (marcrasi) [reply Examples/Pose3SLAMG2O/main.swift:49]: This is the most popular swift argument parsing library: https://github.com/apple/swift-argument-parser

Comment (ProfFan) [reply Examples/Pose3SLAMG2O/main.swift]: done

File: Sources/SwiftFusion/Datasets/G2OReader.swift [hunk 2]
```diff
@@ -79,7 +79,7 @@ public enum G2OReader {
           assert(typedID.perTypeID == id)
           variableId.append(typedID)
         case .measurement(frameIndex: let id1, measuredIndex: let id2, pose: let difference):
-          graph.store(NewBetweenFactor3(TypedID(id1), TypedID(id2), difference))
+          graph.store(NewBetweenFactorAlternative3(TypedID(id1), TypedID(id2), difference))
```
Comment (dellaert) [inline Sources/SwiftFusion/Datasets/G2OReader.swift:82]: How about not hardcoding this ? Add a command line argument and an option here...

File: Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import PenguinStructures
+
+/// A factor that specifies a difference between two poses.
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift:17]: This says nothing about Frobenius norm

File: Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  public init(_ startId: TypedID<Pose3, Int>, _ endId: TypedID<Pose3, Int>, _ difference: Pose3) {
+    self.edges = Tuple2(startId, endId)
+    self.difference = difference
+  }
+
+  public typealias ErrorVector = Vector12
+  public func errorVector(_ start: Pose3, _ end: Pose3) -> ErrorVector {
+    let actualMotion = between(start, end)
+    let R = actualMotion.coordinate.rot.coordinate.R + (-1) * difference.rot.coordinate.R
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift:39]: - does not work??

File: Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  public let edges: Variables.Indices
+  public let difference: Pose3
+
+  public init(_ startId: TypedID<Pose3, Int>, _ endId: TypedID<Pose3, Int>, _ difference: Pose3) {
+    self.edges = Tuple2(startId, endId)
+    self.difference = difference
+  }
+
+  public typealias ErrorVector = Vector12
+  public func errorVector(_ start: Pose3, _ end: Pose3) -> ErrorVector {
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift:37]: Don't see a unit test for this

**Comment by dellaert** [comment]
Another note: the lambda behaves strangely. Would be nice to see the same plots on GTSAM

File: Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift (hunk 4 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift:39]: Yes, this is a protocol constraint issue

File: Examples/Pose3SLAMG2O/main.swift (hunk 1 above)
Comment (ProfFan) [reply Examples/Pose3SLAMG2O/main.swift]: @marcrasi Oh I didn't see your comment... I end up using SPMUtility but we can switch at anytime :)

File: Sources/SwiftFusion/Datasets/G2OReader.swift (hunk 2 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Datasets/G2OReader.swift]: Done

File: Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift (hunk 3 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift]: done

File: Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift (hunk 5 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Inference/NewBetweenFactorAlternative.swift:33]: done

**Comment by ProfFan** [comment]
@dellaert Yeah your instinct looks very legit... for example on grid3D we have this clear indication of convergence, as lambda jumps down to 1e-10 for chordal and stays very high for old BetweenFactor:
image

Need more investigation...

**Comment by ProfFan** [comment]
I think we should merge this before our meeting @dellaert

**Comment by dellaert** [comment]
> This is pretty opaque to me. How about adding a readme file on what you learned about te flow of this? With references to source material?

add the readme and we can merge ;-)

**Pull request merged**